### ISSUE TITLE
Ellipsemodel.estimate: some data points lead to scaling issues

### ISSUE THREAD

**Issue by testor2897** [scikit-image/scikit-image#7337]
### Description:

Hello,

Using Ellipsemodel.estimate you sometimes face scaling problems.
Means, that some data points for the estimation lead to mathematical instable calculations that end up in result = None.
In order to avoid this problem, data points can be transformed into centroid and scaled data points.
So I suggest to scale the data points before looking for the solution.
For testing I give a comparison example between scaled and unscaled data points.

Greetings,

Sebastian

### Way to reproduce:

```python
import numpy as np
from skimage.measure import EllipseModel

# scale measuring data for optimized calculations
# move data central around average values
# scale numbers
# return scaled data and scales
def scalePoints(x,y,enableMove=True,enableScale=True):

    if enableMove:
      # calculate average of x and y
      avgx = np.mean(x)
      avgy = np.mean(y)
    else:
      avgx = avgy = 0

    if enableScale:
      # calculating scale factors
      sfx = (np.max(x) - np.min(x))/2
      sfy = (np.max(y) - np.min(y))/2
      # uniform scale factor = max(sfx, sfy)
      # is easier for unscaling
      sfx = sfy = max(sfx, sfy)
      
    else:
      sfx = sfy = 1

    # creating scaled and centroid data points
    # in order to keep calculation results small
    x = (x - avgx)/sfx
    y = (y - avgy)/sfy
    return [x, y, avgx, avgy, sfx, sfy]


def fitEllipse(x,y, useScale=False):
    if useScale:
      # creating scaled and centroid data points
      # in order to keep calculation results small
      x, y, avgx, avgy, sfx, sfy = scalePoints(x,y)

    points = np.hstack([x, y])

    ell = EllipseModel()
    ell.estimate(points)
    if ell.params is None:
      return ell.params
      
    xc, yc, a, b, phi = ell.params

    if useScale:
      # unscale results
      xc = xc * sfx + avgx
      yc = yc * sfy + avgy
      a = a * sfx
      b = b * sfy
    
    return [xc, yc, a, b, phi]


titleSample = "pretty linear data"
xSample = np.array([3.0751, 2.9036, 2.7675, 2.6505, 2.5423, 2.4352])
ySample = np.array([3.2164, 3.133, 3.0626, 2.9991, 2.9378, 2.8748])

print("Data Points\nx:\n", xSample, "\ny:\n", ySample, "\n")
# transpose coordinates an make it 2 dimensional
xSample = xSample[:, np.newaxis]
ySample = ySample[:, np.newaxis]

# scale points for fit
print("\nResult of estimate with scaled points:\n[xc, yc, a, b, phi]")
para = fitEllipse(xSample,ySample, True)
print(para)

# do not scale points for fit
print("\nResult of estimate without scaling:\n[xc, yc, a, b, phi]")
para = fitEllipse(xSample,ySample, False)
print(para)

```

### Version information:

```Shell
3.7.9 (tags/v3.7.9:13c94747c7, Aug 17 2020, 18:58:18) [MSC v.1900 64 bit (AMD64)]
Windows-10-10.0.19041-SP0
scikit-image version: 0.19.3
numpy version: 1.21.6
```

**Comment by mkcor** [comment]
Hi @testor2897,

Thanks for reporting. In order to make your issue easier to investigate, would you mind replacing `from math import *` with an explicit `import math` and, perhaps, attach some plots to visualize data points?

**Comment by testor2897** [comment]
Hi,

I removed `import math` from the example as it is unnecessary, sorry for conufsion.

In order to make a visualization I produced a matplot-graph:

```python
from pylab import *
from matplotlib.ticker import (MultipleLocator, FormatStrFormatter, AutoMinorLocator)

import numpy as np

try:
  from skimage.measure import EllipseModel
  scikitloaded = True
except:
  scikitloaded = False
  print("\nscikit not installed!\n")


# scale measuring data for optimized calculations
# move data central around average values
# scale numbers to <2
# return scaled data and scales
def scalePoints(x,y,enableMove=True,enableScale=True):

    if enableMove:
      # calculate average of x and y
      avgx = np.mean(x)
      avgy = np.mean(y)
    else:
      avgx = avgy = 0

    if enableScale:
      # calculating scale factors
      sfx = (np.max(x) - np.min(x))/2
      sfy = (np.max(y) - np.min(y))/2
      # uniform scale factor = max(sfx, sfy)
      # is easier for unscaling
      sfx = sfy = max(sfx, sfy)
      
    else:
      sfx = sfy = 1

    # creating scaled and centroid data points
    # in order to keep calculation results small
    x = (x - avgx)/sfx
    y = (y - avgy)/sfy
    return [x, y, avgx, avgy, sfx, sfy]


# https://stackoverflow.com/questions/39693869/fitting-an-ellipse-to-a-set-of-data-points-in-python
def fitEllipse(x,y, useScale=False):
    if useScale:
      # creating scaled and centroid data points
      # in order to keep calculation results small
      x, y, avgx, avgy, sfx, sfy = scalePoints(x,y)

    points = np.hstack([x, y])

    ell = EllipseModel()
    ell.estimate(points)
    if ell.params is None:
      return ell.params
      
    xc, yc, a, b, phi = ell.params

    if useScale:
      # unscale results
      xc = xc * sfx + avgx
      yc = yc * sfy + avgy
      a = a * sfx
      b = b * sfy
    
    return [xc, yc, a, b, phi]


def get_ellipse_pts(params, npts=100, tmin=0, tmax=2*np.pi):
    """
    Return npts points on the ellipse described by the params = x0, y0, ap,
    bp, e, phi for values of the parametric variable t between tmin and tmax.

    """

    x0, y0, ap, bp, phi = params
    # A grid of the parametric variable, t.
    t = np.linspace(tmin, tmax, npts)
    x = x0 + ap * np.cos(t) * np.cos(phi) - bp * np.sin(t) * np.sin(phi)
    y = y0 + ap * np.cos(t) * np.sin(phi) + bp * np.sin(t) * np.cos(phi)
    return x, y


titleSample = "pretty linear data"
xSample = np.array([3.0751, 2.9036, 2.7675, 2.6505, 2.5423, 2.4352])
ySample = np.array([3.2164, 3.133, 3.0626, 2.9991, 2.9378, 2.8748])

print("Data Points\nx:\n", xSample, "\ny:\n", ySample, "\n")
# Transpose Coordinates an make it 2 dimensional
xSample = xSample[:, np.newaxis]
ySample = ySample[:, np.newaxis]

fig, ax = plt.subplots()
plot(xSample, ySample, 'bo', label='data points')
xlabel('x')
ylabel('y')
ax.xaxis.set_minor_locator(AutoMinorLocator())
ax.yaxis.set_minor_locator(AutoMinorLocator())

# Prepare Plot
fig.canvas.manager.set_window_title('Fitting of Ellipses with data points')
fig.patch.set_facecolor('white')



if scikitloaded:
    # scale points for fit
    print("\nResult of estimate with scaled points:\n[xc, yc, a, b, phi]")
    para = fitEllipse(xSample,ySample, True)
    print(para)
    xx, yy = get_ellipse_pts(para)
    plot(xx, yy, 'g-' , linewidth = 1, label = "Result of estimate with scaled points")
    xc, yc = para[0:2]
    plot(xc, yc, 'k.')

    # do not scale points for fit
    print("\nResult of estimate without scaling:\n[xc, yc, a, b, phi]")
    para = fitEllipse(xSample,ySample, False)
    print(para)
    if para is not None:
        xx, yy = get_ellipse_pts(para)
        plot(xx, yy, 'r-' , linewidth = 1, label = "Result of estimate without scaling")
        xc, yc = para[0:2]
        plot(xc, yc, 'b.')

    # set general plot parameter
    legend(loc='lower right', labelcolor='linecolor', facecolor="darkgray")
    plt.show()
```

The points and results in console are:
![image](https://github.com/scikit-image/scikit-image/assets/89394187/b4af3fa5-bf21-43da-8085-cf061b4b53cf)

There is the result None (no result or error) for the test without scaling.

Matplotlib (extended example) shows the following:
![image](https://github.com/scikit-image/scikit-image/assets/89394187/d985b2a1-71c2-4000-9681-fb3852e734f0)

As there is no result for the test without scaling the example won't plot a graph.

**Comment by mkcor** [comment]
Hi @testor2897,

Thank you. I think I get it. Basically, you recommend making the data values non-dimensional (falling in the interval [-1, 1]) before fitting the model. The result can be re-dimensionalized afterwards, indeed. As a physicist by training, I can't say no to that!

Would you be able/available to submit a PR updating `skimage.measure.EllipseModel`?

**Comment by mkcor** [comment]
Hi Sebastian,

> unfortunately I am not familiar with PR updating (pull requests?) and had little time lately.

Yes, 'PR' means 'pull request' (pardon the abbreviation). I wasn't talking about 'PR updating' actually; I was talking about *submitting a pull request which would update `skimage.measure.EllipseModel`* (sorry that wasn't clear).

> So I extracted the newest version of estimate (part of Class EllipseModel in fit.py) today and tested. It seems that meanwhile the code was updated...

You can always see whether code was updated or not, and when exactly, and by who. Here is the [commit history](https://github.com/scikit-image/scikit-image/commits/main/skimage/measure/fit.py) for `fit.py`, i.e., the history of all changes which affected (at least one line of) this file. Here is the [Git blame](https://github.com/scikit-image/scikit-image/blame/main/skimage/measure/fit.py) for `fit.py`, i.e., the detail of all changes for each line. I'm guessing that the code change you've noticed comes from PR #6703, which was merged on Mar 2, 2023 and landed in version 0.21.

You shared your version information in the first place (scikit-image version: 0.19.3) but I'm taking a look at it only now! :see_no_evil: 

> My new test was positive => parameters were produced: xc, yc, a, b, phi = Ellipse.params and even a > b that is nice. When will the new version be released?

We just released version 0.23.1, as you can see from [PyPI](https://pypi.org/project/scikit-image/#history) or the [release page](https://github.com/scikit-image/scikit-image/releases) of the repo.

Best,
Marianne
